### PR TITLE
Separating scenario logic from the VM implementation

### PR DESCRIPTION
**Pull request by andrei-marinica**
It is important to completely separate the logic of scenarios from the logic of the VM execution.

Not in scope of this PR: moving the scenario logic to the multiversx-sc-scenario crate. Will come next, I didn't want to also move that much code around.

### CODE REVIEW

File: framework/scenario/src/facade/scenario_world_runner.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        }
+    }
+}
+
+impl StepRunner for ScenarioWorld {
+    fn run_external_steps(&mut self, step: &ExternalStepsStep) {
+        self.for_each_runner_mut(|runner| runner.run_external_steps(step));
+    }
+
+    fn run_set_state_step(&mut self, step: &SetStateStep) {
```
Comment (CostinCarabas) [inline framework/scenario/src/facade/scenario_world_runner.rs:18]: Also, all these steps like `SetStateStep` are implemented in `vm`.

**Review by CostinCarabas — APPROVED** [review]

File: framework/scenario/src/facade/scenario_world.rs [hunk 2]
```diff
@@ -13,40 +20,51 @@ use std::path::{Path, PathBuf};
...
+    pub vm_runner: VmAdapter, // TODO: convert to option at some point
+    pub trace: Option<ScenarioTrace>,
 }
 
 impl ScenarioWorld {
     pub fn new() -> Self {
         ScenarioWorld {
-            blockchain_mock: BlockchainMock::new(),
+            current_dir: std::env::current_dir().unwrap(),
+            vm_runner: VmAdapter::new(),
```
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply framework/scenario/src/facade/scenario_world.rs:32]: Yes, but indirectly. That is by design.

File: framework/scenario/src/facade/scenario_world_runner.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+impl ScenarioWorld {
+    pub fn for_each_runner_mut<F: FnMut(&mut dyn StepRunner)>(&mut self, mut f: F) {
+        f(&mut self.vm_runner);
+        if let Some(trace) = &mut self.trace {
+            f(trace);
+        }
+    }
+}
+
+impl StepRunner for ScenarioWorld {
```
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply framework/scenario/src/facade/scenario_world_runner.rs:13]: In next PR.

File: framework/scenario/src/facade/scenario_world_steps.rs [hunk 4]
```diff
@@ -5,49 +5,58 @@ use crate::{
...
+use multiversx_chain_vm::{
+    multiversx_sc::{codec::PanicErrorHandler, types::ContractCall},
+    scenario::{
+        handler::{StepHandler, StepRunner},
+        model::*,
+    },
+    DebugApi,
+};
 
 impl StepHandler for ScenarioWorld {
```
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply framework/scenario/src/facade/scenario_world_steps.rs:17]: They are not the same. StepHandler defines a builder, StepRunner is backend. Will discuss.

File: vm/src/scenario/executor/sc_query.rs [hunk 5]
```diff
@@ -1,65 +1,22 @@
...
 };
 use multiversx_sc::{
-    codec::{CodecFrom, PanicErrorHandler, TopEncodeMulti},
+    codec::{CodecFrom, PanicErrorHandler},
     types::ContractCall,
 };
 
 use super::check_tx_output;
 
 impl BlockchainMock {
```
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply vm/src/scenario/executor/sc_query.rs:16]: Cleanup will happen later.

File: framework/scenario/src/facade/scenario_world_runner.rs (hunk 1 above)
Comment (andrei-marinica) [reply framework/scenario/src/facade/scenario_world_runner.rs:18]: This PR was not about moving, just yet.

**Pull request merged**